docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,214 @@
+
+### Crossplane
+
+Another alternative is [Crossplane](https://www.crossplane.io/), a CNCF project which is a set of Kubernetes controllers and Custom Resource Definitions like ASO that define and operate on resources for Azure and other cloud providers. A number of factors indicate that ASO may be a better fit for CAPZ:
+- CRD definitions generated from the Azure Terraform provider maintained by
+  HashiCorp, which may not keep up with changes as fast as the Azure API swagger
+  specification from which ASO generates its definitions
+- Still using the [deprecated Azure Go SDK](https://github.com/crossplane/crossplane/blob/08f11debb66cdf10ff0ddd14f1ee23aa9b4489b1/go.mod#L46-L55), no apparent plans to update
+- [Open issue](https://github.com/upbound/provider-azure/issues/94) to support workload identity, which ASO already handles
+
+## Upgrade Strategy
+
+There are two main upgrade paths CAPZ needs to consider: the first between a version of CAPZ that does not use ASO to one that does, and the second between two versions of CAPZ that use different versions of ASO.
+
+During the initial adoption of ASO, CAPZ's ASO-based [reconciliation flow](#reconciliation) will automatically adopt resources previously managed by CAPZ based on the `sigs.k8s.io_cluster-api-provider-azure_cluster_*` tag on the Azure resource.
+
+After that, a CAPZ upgrade may include a bump in the API version used for some ASO resource types. ASO will handle those changes transparently from CAPZ's perspective with its own conversion webhooks. The limit on the size of Kubernetes resources will likely force ASO to remove certain API versions of resources in CRDs in the future. When a version is deprecated in ASO, CAPZ will have to upgrade its use of the resource to a newer API version.
+
+Separate from API version upgrades of resources, upgrades to ASO itself may be included in new versions of CAPZ. [ASO's documentation](https://azure.github.io/azure-service-operator/introduction/upgrading/) recommends upgrading using the same mechanism used to initially install. At the same time when CAPZ installs ASO, it will also upgrade an existing ASO it installed if necessary.
+
+## Additional Details
+
+### Test Plan
+
+Adoption of ASO to manage Azure infrastructure is at the core of CAPZ's functionality. As such, changes made to integrate ASO must be thoroughly tested.
+
+e2e tests will give the best indication if any regressions are introduced. Because no user-facing changes to CAPZ are expected as a result of integrating with ASO, all of the existing e2e tests will not need any changes to continue providing the same value they do already and do not need to compensate for the transition period where ASO is only partially integrated. New tests or additions to the test framework may include the following:
+- a checkpoint when installing CAPZ in the framework that ASO is also installed and healthy
```

Comment:
Will this test also ensure that if there are any existing copies of ASO, that the version is compatible with CAPZ?

Reply:
I suppose it could. I think the version of ASO is less important than the API versions of the ASO resources CAPZ uses, and that should fail pretty reasonably during an existing test if something's not right there. And I think that would be a lot harder to check up front.